File: rs_ai/backend/services/news_service.py
import os
import requests
from newsapi import NewsApiClient
from .cache_service import get_cached_data, set_cached_data

# Initialize NewsAPI client
news_api_key = os.environ.get('NEWS_API_KEY')
newsapi = NewsApiClient(api_key=news_api_key)

def get_news_articles(country_code, query):
    """
    Get news articles related to a country and query
    
    Args:
        country_code: ISO country code
        query: Search query
        
    Returns:
        List of news articles
    """
    cache_key = f"news_{country_code}_{query}"
    cached_data = get_cached_data(cache_key, 3600)  # Cache for 1 hour
    
    if cached_data:
        return cached_data
    
    try:
        # Get country name for better search results
        country_name = get_country_name(country_code)
        
        # Build search query
        search_query = f"{country_name} {query}"
        
        # Get articles from NewsAPI
        response = newsapi.get_everything(
            q=search_query,
            language='en',
            sort_by='relevancy',
            page=1,
            page_size=10
        )
        
        articles = []
        
        for article in response.get('articles', []):
            # Generate tags from title and description
            tags = generate_tags(
                article.get('title', ''),
                article.get('description', ''),
                query
            )
            
            # Format article data
            articles.append({
                'title': article.get('title'),
                'source': article.get('source', {}).get('name'),
                'date': article.get('publishedAt')[:10],  # YYYY-MM-DD
                'description': article.get('description'),
                'url': article.get('url'),
                'imageUrl': article.get('urlToImage'),
                'tags': tags
            })
        
        # Cache the result
        set_cached_data(cache_key, articles, 3600)  # Cache for 1 hour
        
        return articles
    except Exception as e:
        print(f"Error fetching news articles: {str(e)}")
        return []

def analyze_news_sentiment(articles):
    """
    Analyze sentiment of news articles
    
    Args:
        articles: List of news articles
        
    Returns:
        Sentiment analysis results
    """
    # In a real application, you would use a more sophisticated
    # sentiment analysis approach, possibly using an AI service
    
    # For this example, we'll use a simple approach
    positive_keywords = [
        'growth', 'profit', 'success', 'positive', 'increase',
        'improve', 'gain', 'benefit', 'opportunity', 'innovation'
    ]
    
    negative_keywords = [
        'decline', 'loss', 'fail', 'negative', 'decrease',
        'risk', 'threat', 'crisis', 'problem', 'debt'
    ]
    
    # Initialize counters
    positive_count = 0
    negative_count = 0
    neutral_count = 0
    
    # Count occurences of positive and negative keywords
    for article in articles:
        title = article.get('title', '').lower()
        description = article.get('description', '').lower()
        content = title + ' ' + description
        
        article_positive = 0
        article_negative = 0
        
        for keyword in positive_keywords:
            if keyword in content:
                article_positive += 1
        
        for keyword in negative_keywords:
            if keyword in content:
                article_negative += 1
        
        # Classify article sentiment
        if article_positive > article_negative:
            positive_count += 1
        elif article_negative > article_positive:
            negative_count += 1
        else:
            neutral_count += 1
    
    # Calculate percentages
    total = max(1, len(articles))  # Avoid division by zero
    positive_percent = round((positive_count / total) * 100)
    negative_percent = round((negative_count / total) * 100)
    neutral_percent = round((neutral_count / total) * 100)
    
    # Generate summary
    if positive_percent > negative_percent and positive_percent > neutral_percent:
        summary = "Generally positive sentiment in recent news coverage"
    elif negative_percent > positive_percent and negative_percent > neutral_percent:
        summary = "Predominantly negative sentiment in recent news coverage"
    else:
        summary = "Mostly neutral sentiment in recent news coverage"
    
    return {
        'positive': positive_percent,
        'negative': negative_percent,
        'neutral': neutral_percent,
        'summary': summary
    }

def generate_tags(title, description, query):
    """
    Generate tags from article title and description
    
    Args:
        title: Article title
        description: Article description
        query: Original search query
        
    Returns:
        List of tags
    """
    # Combine title and description
    text = (title + " " + description).lower()
    
    # Potential tag categories
    economic_tags = ['economy', 'gdp', 'growth', 'inflation', 'recession', 'market', 'trade']
    political_tags = ['government', 'policy', 'election', 'reform', 'regulation', 'law']
    business_tags = ['business', 'company', 'investment', 'startup', 'corporation', 'industry']
    social_tags = ['society', 'health', 'education', 'culture', 'employment', 'welfare']
    
    # Find matches
    tags = []
    
    # Add query as a tag if it's not too long
    if len(query) < 20:
        tags.append(query)
    
    # Check for tag matches
    for tag in economic_tags + political_tags + business_tags + social_tags:
        if tag in text and tag not in tags and len(tags) < 5:
            tags.append(tag)
    
    return tags

def get_country_name(country_code):
    """
    Get country name from country code
    
    Args:
        country_code: ISO country code
        
    Returns:
        Country name
    """
    # Map of common country codes to names
    country_map = {
        'USA': 'United States',
        'GBR': 'United Kingdom',
        'DEU': 'Germany',
        'FRA': 'France',
        'JPN': 'Japan',
        'CHN': 'China',
        'IND': 'India',
        'BRA': 'Brazil',
        'CAN': 'Canada',
        'AUS': 'Australia',
        'RUS': 'Russia',
        'KOR': 'South Korea',
        'ITA': 'Italy',
        'ESP': 'Spain',
        'MEX': 'Mexico'
    }
    
    return country_map.get(country_code, country_code)

File: server/services/newsService.ts
import fetch from 'node-fetch';
import { NewsItem, NewsSentiment } from '@shared/schema';
import { getMistralInsights } from './mistralService';

// Get API key from environment variable
const NEWS_API_KEY = process.env.NEWS_API_KEY || "";

// Get news articles for a country
export async function getNewsArticles(countryCode: string, query: string): Promise<NewsItem[]> {
  try {
    if (!NEWS_API_KEY) {
      throw new Error("News API key is not configured");
    }
    
    // Get country name from the World Bank API
    const countryResponse = await fetch(`https://api.worldbank.org/v2/country/${countryCode}?format=json`);
    const countryData = await countryResponse.json();
    const countryName = countryData[1][0].name;
    
    // Prepare search query
    const searchQuery = `${countryName} ${query}`;
    
    // Call News API
    const response = await fetch(
      `https://newsapi.org/v2/everything?q=${encodeURIComponent(searchQuery)}&sortBy=relevancy&language=en&pageSize=10`,
      {
        headers: {
          'X-Api-Key': NEWS_API_KEY
        }
      }
    );
    
    if (!response.ok) {
      throw new Error(`News API error: ${response.status}`);
    }
    
    const data = await response.json();
    
    // When no articles were found
    if (!data.articles || data.articles.length === 0) {
      return [];
    }
    
    // Format articles for our app
    return data.articles.map((article: any) => {
      // Generate tags based on article content
      const tags = generateTags(article.title, article.description, query);
      
      // Format the date
      const publishedDate = new Date(article.publishedAt);
      const now = new Date();
      const diffDays = Math.floor((now.getTime() - publishedDate.getTime()) / (1000 * 60 * 60 * 24));
      
      let formattedDate;
      if (diffDays === 0) {
        formattedDate = "Today";
      } else if (diffDays === 1) {
        formattedDate = "Yesterday";
      } else if (diffDays < 7) {
        formattedDate = `${diffDays} days ago`;
      } else if (diffDays < 30) {
        formattedDate = `${Math.floor(diffDays / 7)} weeks ago`;
      } else {
        formattedDate = publishedDate.toLocaleDateString();
      }
      
      return {
        title: article.title,
        source: article.source.name,
        date: formattedDate,
        description: article.description || "No description available",
        url: article.url,
        imageUrl: article.urlToImage,
        tags
      };
    });
  } catch (error) {
    console.error("Error fetching news articles:", error);
    throw error;
  }
}

// Generate tags based on article content
function generateTags(title: string, description: string, query: string): string[] {
  const content = `${title} ${description}`.toLowerCase();
  const tags: string[] = [];
  
  // Economic keywords
  if (content.includes('economy') || content.includes('economic') || content.includes('gdp') || 
      content.includes('growth') || content.includes('recession')) {
    tags.push('Economy');
  }
  
  // Technology keywords
  if (content.includes('tech') || content.includes('technology') || content.includes('digital') || 
      content.includes('innovation') || content.includes('startup')) {
    tags.push('Technology');
  }
  
  // Policy keywords
  if (content.includes('policy') || content.includes('regulation') || content.includes('law') || 
      content.includes('government') || content.includes('reform')) {
    tags.push('Monetary Policy');
  }
  
  // Real estate keywords
  if (content.includes('real estate') || content.includes('housing') || content.includes('property') || 
      content.includes('mortgage') || content.includes('home')) {
    tags.push('Real Estate');
  }
  
  // Employment keywords
  if (content.includes('job') || content.includes('employment') || content.includes('unemployment') || 
      content.includes('labor') || content.includes('workforce')) {
    tags.push('Employment');
  }
  
  // Trade keywords
  if (content.includes('trade') || content.includes('export') || content.includes('import') || 
      content.includes('tariff') || content.includes('global')) {
    tags.push('Trade');
  }
  
  // If no tags were found, add a general one based on the query
  if (tags.length === 0) {
    // Extract main topic from query
    const queryWords = query.split(' ');
    const mainTopic = queryWords.find(word => word.length > 3) || 'General';
    tags.push(mainTopic.charAt(0).toUpperCase() + mainTopic.slice(1));
  }
  
  return tags;
}

// Analyze news sentiment
export async function analyzeNewsSentiment(articles: NewsItem[]): Promise<NewsSentiment> {
  try {
    if (articles.length === 0) {
      return {
        neutral: 60,
        positive: 20,
        negative: 20,
        summary: "No recent news articles were found for analysis."
      };
    }
    
    // This would normally use NLP or the Mistral API for sentiment analysis
    // For this implementation, we'll use a simple rule-based approach
    
    let positive = 0;
    let negative = 0;
    let neutral = 0;
    
    // Positive and negative keywords
    const positiveWords = ['growth', 'increase', 'gain', 'improved', 'positive', 'success', 'recovery', 'up', 
                         'rising', 'opportunity', 'profit', 'progress', 'strong', 'benefit', 'advance'];
    
    const negativeWords = ['decline', 'decrease', 'loss', 'crisis', 'concern', 'negative', 'fail', 'down', 
                          'falling', 'problem', 'deficit', 'weak', 'risk', 'threat', 'challenge'];
    
    // Count sentiment in each article
    articles.forEach(article => {
      const content = `${article.title} ${article.description}`.toLowerCase();
      
      let positiveScore = positiveWords.reduce((score, word) => {
        return score + (content.includes(word) ? 1 : 0);
      }, 0);
      
      let negativeScore = negativeWords.reduce((score, word) => {
        return score + (content.includes(word) ? 1 : 0);
      }, 0);
      
      if (positiveScore > negativeScore) {
        positive++;
      } else if (negativeScore > positiveScore) {
        negative++;
      } else {
        neutral++;
      }
    });
    
    // Calculate percentages
    const total = articles.length;
    const positivePercent = (positive / total) * 100;
    const negativePercent = (negative / total) * 100;
    const neutralPercent = (neutral / total) * 100;
    
    // Generate summary
    let sentimentTrend;
    if (positivePercent > 40) {
      sentimentTrend = "predominantly positive";
    } else if (negativePercent > 40) {
      sentimentTrend = "predominantly negative";
    } else {
      sentimentTrend = "mostly neutral";
    }
    
    const topicWords = articles.map(a => a.tags).flat();
    const topicCounts: Record<string, number> = {};
    
    topicWords.forEach(topic => {
      topicCounts[topic] = (topicCounts[topic] || 0) + 1;
    });
    
    // Get top 3 topics
    const topTopics = Object.entries(topicCounts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 3)
      .map(entry => entry[0]);
      
    const summary = `Recent news coverage shows ${sentimentTrend} sentiment (${Math.round(neutralPercent)}% neutral, ${Math.round(positivePercent)}% positive, ${Math.round(negativePercent)}% negative). Key topics in the reporting include ${topTopics.join(", ")}.`;
    
    return {
      neutral: neutralPercent,
      positive: positivePercent,
      negative: negativePercent,
      summary
    };
  } catch (error) {
    console.error("Error analyzing news sentiment:", error);
    
    // Return a default sentiment if analysis fails
    return {
      neutral: 60,
      positive: 20,
      negative: 20,
      summary: "Unable to perform sentiment analysis on the available news articles."
    };
  }
}
